src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -1985,183 +1968,174 @@ int16_t LoRaWANNode::parseDownlink(uint8_t* data, size_t* len, uint8_t window, L
-  if(fOptsLen > 0) {
+    // remove all MAC commands except those whose payload can be requested by the user
+    // (which are LinkCheck and DeviceTime)  
     LoRaWANNode::clearMacCommands(fOpts, &fOptsLen, RADIOLIB_LORAWAN_DOWNLINK);
     memcpy(this->fOptsDown, fOpts, fOptsLen);
-  }
-  this->fOptsDownLen = fOptsLen;
-  
-  // if fOptsLen for the next uplink is larger than can be piggybacked onto an uplink, send separate uplink
-  if(fOptsReLen > RADIOLIB_LORAWAN_FHDR_FOPTS_MAX_LEN) {
+    this->fOptsDownLen = fOptsLen;
 
-    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Uplink MAC-only payload (%d bytes):", fOptsReLen);
-    RADIOLIB_DEBUG_PROTOCOL_HEXDUMP(fOptsRe, fOptsReLen);
+    // if fOptsLen for the next uplink is larger than can be piggybacked onto an uplink, send separate uplink
+    if(fOptsReLen > RADIOLIB_LORAWAN_FHDR_FOPTS_MAX_LEN) {
 
-    this->isMACPayload = true;
-    // temporarily lift dutyCycle restrictions to allow immediate MAC response
-    bool prevDC = this->dutyCycleEnabled;
-    this->dutyCycleEnabled = false;
+      RADIOLIB_DEBUG_PROTOCOL_PRINTLN("! Sending MAC-only uplink (%d bytes):", fOptsReLen);
+      RADIOLIB_DEBUG_PROTOCOL_HEXDUMP(fOptsRe, fOptsReLen);
 
-    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Sending MAC-only uplink .. ");
+      this->isMACPayload = true;
 
-    this->sendReceive(fOptsRe, fOptsReLen, RADIOLIB_LORAWAN_FPORT_MAC_COMMAND);
+      // temporarily lift dutyCycle restrictions to allow immediate MAC response
+      bool prevDC = this->dutyCycleEnabled;
+      this->dutyCycleEnabled = false;
+      this->sendReceive(fOptsRe, fOptsReLen, RADIOLIB_LORAWAN_FPORT_MAC_COMMAND);
+      this->dutyCycleEnabled = prevDC;
 
-    this->dutyCycleEnabled = prevDC;
+    } else { // fOptsReLen <= 15
+      memcpy(this->fOptsUp, fOptsRe, fOptsReLen);
+      this->fOptsUpLen = fOptsReLen;
+    }
+  }
 
-  } else { // fOptsReLen <= 15
-    memcpy(this->fOptsUp, fOptsRe, fOptsReLen);
-    this->fOptsUpLen = fOptsReLen;
+  // figure out which key to use to decrypt the payload
+  uint8_t* encKey = this->appSKey;
+  if(this->multicast && window == RADIOLIB_LORAWAN_RX_BC) {
+    encKey = this->mcAppSKey;
+  }
+  for(int id = 0; id < RADIOLIB_LORAWAN_NUM_SUPPORTED_PACKAGES; id++) {
+    if(this->packages[id].enabled && fPort == this->packages[id].packFPort) {
+      encKey = this->packages[id].isAppPack ? this->appSKey : this->nwkSEncKey;
+      break;
+    }
+  }
+
+  // decrypt the frame payload
+  // by default, the data and length are user-accessible
+  processAES(&downlinkMsg[RADIOLIB_LORAWAN_FRAME_PAYLOAD_POS(fOptsLen)], payLen, encKey, data, addr, devFCnt32, RADIOLIB_LORAWAN_DOWNLINK, 0x00, true);
+  *len = payLen;
+
+  // however, if this frame belongs to a package, redirect instead and 'hide' contents from the user
+  // just to be sure that it doesn't get re-interpreted...
+  for(int id = 0; id < RADIOLIB_LORAWAN_NUM_SUPPORTED_PACKAGES; id++) {
+    if(this->packages[id].enabled && fPort == this->packages[id].packFPort) {
```

Comment:
My paranoia of nullpointer dereference is showing, but I would add a check here to make sure `this->packages[id].callback != NULL`. Maybe a simple function `isPackage` to ensure the registered package is complete.